robottelo/hosts.py

```diff
@@ -1806,14 +1807,15 @@ def get_caller():
                 if frame.function.startswith('test_'):
                     return frame.function
 
-        return Session(
+        self._ui_session = Session(
             session_name=testname or get_caller(),
             user=user or settings.server.admin_username,
             password=password or settings.server.admin_password,
             url=url,
             hostname=self.hostname,
             login=login,
         )
+        return self._ui_session
```

Comment:
This is dangerous. Let me explain why.
Let's take a look at `tests/foreman/ui/test_ldap_authentication.py::test_positive_update_external_roles`
https://github.com/SatelliteQE/robottelo/blob/689178e8485581db4b0cf576020773b2ae84e0e0/tests/foreman/ui/test_ldap_authentication.py#L328-L349

Here we see that this test has two nested UI sessions. If these would be converted to use the `ui_session` from `Satellite` class, it would look like this:
```python
    with target_sat.ui_session() as session:
        session.usergroup.create(
            {
                'usergroup.name': ldap_usergroup_name,
                'roles.resources.assigned': [foreman_role.name],
                'external_groups.name': EXTERNAL_GROUP_NAME,
                'external_groups.auth_source': auth_source_name,
            }
        )
        assert session.usergroup.search(ldap_usergroup_name)[0]['Name'] == ldap_usergroup_name
        with target_sat.ui_session(
            test_name, ldap_data['ldap_user_name'], ldap_data['ldap_user_passwd']
        ) as ldapsession:
            with pytest.raises(NavigationTriesExceeded):
                ldapsession.architecture.search('')
            ldapsession.location.create({'name': location_name})
            location = entities.Location().search(query={'search': f'name="{location_name}"'})[0]
            assert location.name == location_name
        session.usergroup.update(
            ldap_usergroup_name, {'roles.resources.assigned': [katello_role.name]}
        )
        session.usergroup.refresh_external_group(ldap_usergroup_name, EXTERNAL_GROUP_NAME)
```

The `ui_session` function call in the inner session (`ldapsession`) rewrites the contents of `target_sat._ui_session`. Once we step out of the inner context manager (UI session), `target_sat` will still have assigned the `ldapsession` in its `_ui_session` attribute.

What is dangerous about it? If we make use of any function from the `ui_factory` in the master session after the inner/sub session was closed, `ui_factory` will try to run the factory function on the inner/sub session instead of on the master/primary/admin session.
 
```python
    with target_sat.ui_session() as session:
        session.usergroup.create(
            {
                'usergroup.name': ldap_usergroup_name,
                'roles.resources.assigned': [foreman_role.name],
                'external_groups.name': EXTERNAL_GROUP_NAME,
                'external_groups.auth_source': auth_source_name,
            }
        )
        assert session.usergroup.search(ldap_usergroup_name)[0]['Name'] == ldap_usergroup_name
        with target_sat.ui_session(
            test_name, ldap_data['ldap_user_name'], ldap_data['ldap_user_passwd']
        ) as ldapsession:
            with pytest.raises(NavigationTriesExceeded):
                ldapsession.architecture.search('')
            ldapsession.location.create({'name': location_name})
            location = entities.Location().search(query={'search': f'name="{location_name}"'})[0]
            assert location.name == location_name

        kwargs_for_fake_host = {}
        target_sat.ui_factory.create_fake_host(**kwargs_for_fake_host)  # <--- We are out of the context manager, but ui_factory gets the ldapsession instead of the "master" session.
```

Reply:
After a discussion on Slack, we decided to go for passing down the `session` parameter to the `ui_factory` function , this would then be used in UIFactory methods to interact with UI.

This saves us from saving and disposing of the session object in the Session context manager.